File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/Views/ActionsMenuView/MenuView/ActionsMenuView.m
//
//  ActionsMenuView.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 15.08.2021.
//  Copyright © 2021 QuickBlox Team. All rights reserved.
//

#import "ActionsMenuView.h"
#import "MenuActionCell.h"
#import "UIView+Videochat.h"

@interface ActionsMenuView () <UITableViewDataSource>
//MARK: - IBOutlets
@property (weak, nonatomic) IBOutlet UIView *containerView;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIButton *cancelButton;

//MARK: - Properties
@property (nonatomic, strong) NSMutableArray<MenuAction *> *actions;
@end

@implementation ActionsMenuView
//MARK: - Life Cycle
- (void)layoutSubviews {
    [super layoutSubviews];
    
    [self setupViews];
}

//MARK: - Actions
- (IBAction)tapCancelButton:(UIButton *)sender {
    [self removeFromSuperview];
}

//MARK - Setup
- (void)setupViews {
    UINib *nibMenuCell = [UINib nibWithNibName:kMenuActionCellIdentifier bundle:nil];
    [self.tableView registerNib:nibMenuCell forCellReuseIdentifier:kMenuActionCellIdentifier];
    self.tableView.userInteractionEnabled = YES;
    
    CGFloat heightConstant = self.tableView.contentSize.height;
    self.containerView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.containerView.topAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.topAnchor constant: 32.0f].active = YES;
    [self.containerView.rightAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.rightAnchor constant: -22.0f].active = YES;
    [self.containerView.widthAnchor constraintEqualToConstant: 200.0f].active = YES;
    [self.containerView.heightAnchor constraintEqualToConstant: heightConstant].active = YES;
    [self.containerView addShadow:[UIColor colorWithRed:0.78f green:0.81f blue:0.85f alpha:1.0f]];
    [self.containerView setRoundViewWithCornerRadius:14.0f];
    
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.tableView.topAnchor constraintEqualToAnchor:self.containerView.topAnchor constant: 3.0f].active = YES;
    [self.tableView.leftAnchor constraintEqualToAnchor:self.containerView.leftAnchor].active = YES;
    [self.tableView.bottomAnchor constraintEqualToAnchor:self.containerView.bottomAnchor constant: -3.0f].active = YES;
    [self.tableView.rightAnchor constraintEqualToAnchor:self.containerView.rightAnchor].active = YES;
    
    [self.tableView reloadData];
}

//MARK: - Public Methods
- (void)addAction:(MenuAction *)action {
    if (!self.actions) {
        self.actions = [NSMutableArray array];
    }
    [self.actions insertObject:action atIndex:0];
}

#pragma mark - Table view data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.actions.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    MenuActionCell *cell = [tableView dequeueReusableCellWithIdentifier:kMenuActionCellIdentifier];
    MenuAction *menuAction = self.actions[indexPath.row];
    cell.actionLabel.text = menuAction.title;
    cell.accessoryType = menuAction.isSelected ? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    MenuAction *menuAction = self.actions[indexPath.row];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self removeFromSuperview] ;
        if (menuAction.handler) {
            menuAction.handler(menuAction.action);
        }
    });
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 44.0f;
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Users/Views/MenuView/ActionsMenuView.swift
//
//  ActionsMenuView.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 14.07.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit

struct MenuConstant {
    static let heightForRow: CGFloat = 44.0
    static let statsInfoWidth: CGFloat = 200.0
}

class ActionsMenuView: UIView {
    //MARK: - IBOutlets
    @IBOutlet weak var tableView: UITableView! {
        didSet {
            tableView.register(UINib(nibName: MenuActionCellConstant.reuseIdentifier, bundle: nil),
                               forCellReuseIdentifier: MenuActionCellConstant.reuseIdentifier)
            tableView.isUserInteractionEnabled = true
        }
    }
    @IBOutlet weak var containerView: UIView!
    @IBOutlet weak var cancelButton: UIButton!

    //MARK: - Properties
    private var actions:[ActionMenu] = []
    
    //MARK: - Life Cycle
    override func layoutSubviews() {
        super.layoutSubviews()
        
        setupViews()
    }
    
    //MARK: - Public Methods
    func addAction(_ action: ActionMenu) {
        actions.append(action)
    }
    
    //MARK: - Actions
    @IBAction func tapCancelButton(_ sender: Any) {
        removeFromSuperview()
    }
    
    //MARK: - Setup
    private func setupViews() {
        let heightConstant = tableView.contentSize.height + 6.0
        containerView.translatesAutoresizingMaskIntoConstraints = false
        containerView.topAnchor.constraint(equalTo: safeAreaLayoutGuide.topAnchor, constant: 32.0).isActive = true
        containerView.rightAnchor.constraint(equalTo: safeAreaLayoutGuide.rightAnchor, constant: -22.0).isActive = true
        containerView.widthAnchor.constraint(equalToConstant: MenuConstant.statsInfoWidth).isActive = true
        containerView.heightAnchor.constraint(equalToConstant: heightConstant).isActive = true
        containerView.addShadowToView(color: #colorLiteral(red: 0.7834715247, green: 0.8073117137, blue: 0.8447045684, alpha: 1))
        containerView.setRoundView(cornerRadius: 14.0)

        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 3.0).isActive = true
        tableView.rightAnchor.constraint(equalTo: containerView.rightAnchor).isActive = true
        tableView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor, constant: -3.0).isActive = true
        tableView.leftAnchor.constraint(equalTo: containerView.leftAnchor).isActive = true
        
        tableView.reloadData()
    }
}

//MARK: -  UITableViewDelegate, UITableViewDataSource
extension ActionsMenuView: UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return actions.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: MenuActionCellConstant.reuseIdentifier, for: indexPath) as? ActionMenuCell else {
            return UITableViewCell()
        }
        
        let menuAction = actions[indexPath.row]
        cell.actionLabel.text = menuAction.title
        cell.accessoryType = menuAction.isSelected == true ? .checkmark : .none
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let menuAction = actions[indexPath.row]
        DispatchQueue.main.async(execute: {
            self.removeFromSuperview()
            menuAction.successHandler?(menuAction.action)
        })
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return MenuConstant.heightForRow
    }
}
